### DISCUSSION TITLE
Selectively run certain configs in vitest workspace?

### DISCUSSION THREAD

**Discussion by jorgebv** [vitest-dev/vitest#3165] [Q&A] [answered]
I am trying to use the new vitest [workspace ](https://vitest.dev/guide/workspace.html) feature. I have gotten most things working, but I haven't been able to figure out how to run just a subset of the tests. I thought I might be able to use the [test filtering CLI option](https://vitest.dev/guide/filtering.html#cli) in order to filter on the folder name, but this seems to not work, possibly because the sub-configs are at a lower level than the folder I am trying to filter on - thus the test name doesn't seem to include the folder-name when executed.

I have assigned the test.name property to each sub-config so they are differentiable on the command-line, and I thought this would be an ideal way to instruct vitest what to run, but I can't find any way to use this property from the command-line.

Is it possible to filter at the config level when using vitest workspaces?

**Comment by sheremet-va** [comment, marked as answer]
Technically test filtering should work 🤔 There is no other way to filter test files currently.

**Reply by jorgebv** [reply]
After some additional experimenting I believe you are right and this is my mistake. I thought that the -t [option](https://vitest.dev/guide/cli.html#options) was equivalent to what is described as [Test Filtering](https://vitest.dev/guide/filtering.html#cli), but it seems they are two different things. When I filter on the folder name with no options, all is working. So in my situation, it seems my problem was trying to use the -t option when I didn't need to.

**Reply by NullVoxPopuli** [reply]
I would love to be able to run `vitest` from a workspace/project directory, and have only those tests run, rather than the whole monorepo

**Reply by sheremet-va** [reply]
> I would love to be able to run `vitest` from a workspace/project directory, and have only those tests run, rather than the whole monorepo

We need to collect tests before filtering them. To collect tests, we need to run the code that defines `test` functions.

**Reply by NullVoxPopuli** [reply]
yeah, that's fine, but don't run the tests / beforeEach stuff. it all receives callbacks, so teh "listing" / "discovery" logic can still work

**Reply by sheremet-va** [reply]
Well, we don't run them if there are no tests to run 🤔

**Reply by satanshiro** [reply]
@sheremet-va I have just configured a monorepo with 507 tests (157 test files) I am trying to also use the workspace feature.
It is a great idea, but currently I cannot run tests based on the configuration. for example i have a vitest.workspace.ts file with the following :
```ts
import { defineWorkspace } from 'vitest/config';

const workspace = defineWorkspace([
  {
    extends: './vitest.config.ts',
    test: {
      name: 'libs',
      include: ['./libs/**/*.spec.ts'],
      setupFiles: [],
    },
  },
  {
    extends: './vitest.config.ts',
    test: {
      name: 'backend-api',
      include: ['./apps/backend-api/**/*.spec.ts'],
      setupFiles: ['./some-setup-file.ts'],
    },
  },]);
export default workspace;
```
And if i run: npx vitest libs 
then the setup file ('some-setup-file.ts') will run regardless that it was configured only for the other workspace.
would make sense to add to the cli something like vitest --workspace \<workspace name\>

**Reply by sheremet-va** [reply]
I think it runs because setupFiles: [] just merges it with the config instead of resetting it

**Reply by satanshiro** [reply]
will test now thanks

**Comment by iamsbro** [comment]
An idea that worked for me was to change the workspace array depending on an environment variable.

`vitest.workspace.ts`
```ts
import { defineWorkspace } from 'vitest/config';

const workspace: Parameters<typeof defineWorkspace>[0] =
    process.env.TEST_SUITE === 'unit'
        ? [
              // unit test config
          ]
        : process.env.TEST_SUITE === 'e2e'
            ? [
                 // e2e test config
              ]
            : []

export default defineWorkspace(workspace);
```

`package.json`
```
"scripts": {
    "test": "TEST_SUITE=unit vitest --run --coverage",
    "e2e": "TEST_SUITE=e2e vitest --run --coverage",
},
```

**Reply by sheremet-va** [reply]
With Vitest 1.0 you can give them names and just pass it down as `--project`:

```ts
export default [
  { test: { name: 'unit' } },
  { test: { name: 'e2e' } },
]
```

```json
"scripts": {
    "test": "vitest --run --coverage --project unit",
    "e2e": "vitest --run --coverage --project e2e",
},
```

If you have several projects, you can also pass them down using this flag:

```json
"scripts": {
    "test": "vitest --run --coverage --project unit1 --project unit2",
},
```

**Reply by iamsbro** [reply]
I'm aware, DX is much better now. Just not able to bump Vitest for my own stuff yet though due to other factors, and figured I'd share for other people in my situation.
